src/CommunityToolkit.Maui.Maps/Handler/Map/MapHandler.Windows.cs

```diff
@@ -366,37 +385,80 @@ void WebViewNavigationCompleted(WebView2 sender, CoreWebView2NavigationCompleted
 
-		if (!string.IsNullOrEmpty(clickedPinWebViewId))
+		if (eventMessage is null)
 		{
-			var clickedPin = VirtualView.Pins.SingleOrDefault(p => (p as Pin)?.Id.ToString().Equals(clickedPinWebViewId) ?? false);
-
-			clickedPin?.SendMarkerClick();
 			return;
 		}
 
-		var clickedInfoWindowWebView = JsonSerializer.Deserialize<InfoWindow>(args.WebMessageAsJson, jsonSerializerOptions);
-		var clickedInfoWindowWebViewId = clickedInfoWindowWebView?.InfoWindowMarkerId;
-
-		if (!string.IsNullOrEmpty(clickedInfoWindowWebViewId))
+		if (!Enum.TryParse<EventIdentifier>(eventMessage.Id, true, out var eventId))
 		{
-			var clickedPin = VirtualView.Pins.SingleOrDefault(p => (p as Pin)?.Id.ToString().Equals(clickedInfoWindowWebViewId) ?? false);
-
-			clickedPin?.SendInfoWindowClick();
 			return;
 		}
 
-		var mapRect = JsonSerializer.Deserialize<Bounds>(args.WebMessageAsJson, jsonSerializerOptions);
-		if (mapRect?.Center is not null)
+		switch (eventId)
 		{
-			VirtualView.VisibleRegion = new MapSpan(new Location(mapRect.Center.Latitude, mapRect.Center.Longitude),
-				mapRect.Height, mapRect.Width);
+			case EventIdentifier.BoundsChanged:
+				var mapRect = JsonSerializer.Deserialize<Bounds>(eventMessage.Payload?.ToString() ?? "", jsonSerializerOptions);
+				if (mapRect?.Center is not null)
+				{
+					VirtualView.VisibleRegion = new MapSpan(new Location(mapRect.Center.Latitude, mapRect.Center.Longitude),
+						mapRect.Height, mapRect.Width);
+				}
+				break;
+			case EventIdentifier.MapClicked:
+				var clickedLocation = JsonSerializer.Deserialize<Location>(eventMessage.Payload?.ToString() ?? "",
+					jsonSerializerOptions);
+				if (clickedLocation is not null)
+				{
+					VirtualView.Clicked(clickedLocation);
+				}
+				break;
+
+			case EventIdentifier.InfoWindowClicked:
+				var clickedInfoWindowWebView = JsonSerializer.Deserialize<InfoWindow>(eventMessage.Payload?.ToString() ?? "",
+					jsonSerializerOptions);
+				var clickedInfoWindowWebViewId = clickedInfoWindowWebView?.InfoWindowMarkerId;
+
+				if (!string.IsNullOrEmpty(clickedInfoWindowWebViewId))
+				{
+					var clickedPin = VirtualView.Pins.SingleOrDefault(p => (p as Pin)?.Id.ToString().Equals(clickedInfoWindowWebViewId) ?? false);
+
+					clickedPin?.SendInfoWindowClick();
+				}
+				break;
+
+			case EventIdentifier.PinClicked:
+				var clickedPinWebView = JsonSerializer.Deserialize<Pin>(eventMessage.Payload?.ToString() ?? "", jsonSerializerOptions);
+				var clickedPinWebViewId = clickedPinWebView?.MarkerId?.ToString();
+
+				if (!string.IsNullOrEmpty(clickedPinWebViewId))
+				{
+					var clickedPin = VirtualView.Pins.SingleOrDefault(p => (p as Pin)?.Id.ToString().Equals(clickedPinWebViewId) ?? false);
+
+					clickedPin?.SendMarkerClick();
+				}
+				break;
 		}
 	}
 
+	class EventMessage
+	{
+		public string Id { get; set; } = string.Empty;
+
+		public object? Payload { get; set; }
```

Comment:
In which case payload is null? Can we mark it as required?

As we always call ToString, can we change the type to string?

Reply:
We could make it required. It’s not really meant to be instantiated by hand though, it should always come from the Bing Map.

In the first commit it was a string type, I guessed that would work. For some reason that throws an exception 😅

Reply:
Maybe we can make it generic because you pass a parameter that is not actually a string (like InfoWindow).

Reply:
Everything is serialized as JSON, so it should be a string I guess? But not 100% sure what you mean. Feel free to do any suggestions or make changes if you have an idea around this.

Reply:
Do you mean dynamic instead of generic?

Reply:
I thought we can put generic type and it can be deserialized automatically. but I see we don't know the target type at this moment.